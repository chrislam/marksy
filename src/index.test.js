import React, {createElement, Component} from 'react';
import Preact from 'preact';
import preactRenderToString from 'preact-render-to-string';
import {renderToString as infernoRenderToString} from 'inferno-server';
import infernoCreateElement from 'inferno-create-element';
import renderer from 'react-test-renderer';
import { transform as babelStandaloneTransform } from 'babel-standalone';
import { transform as babelCoreTransform } from 'babel-core';
import marksy from './'
import marksyComponents from './components'
import Prism from 'prismjs'
import hljs from 'highlight.js/lib/highlight'
import hljsJs from 'highlight.js/lib/languages/javascript'
import hljsXml from 'highlight.js/lib/languages/xml'

hljs.registerLanguage('javascript', hljsJs)
hljs.registerLanguage('xml', hljsXml)

class TestComponent extends Component {
  render () {
    return (
      <div>
        {this.props.children}
      </div>
    )
  }
}

it('should be able to compile text', () => {
  const compile = marksy({createElement});
  const compiled = compile(`hello`);
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile strong text', () => {
  const compile = marksy({createElement});
  const compiled = compile(`hello **there**`);
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile strong text', () => {
  const compile = marksy({createElement});
  const compiled = compile(`hello **there**`);
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile italic text', () => {
  const compile = marksy({createElement});
  const compiled = compile(`hello *there*`);
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile links', () => {
  const compile = marksy({createElement});
  const compiled = compile(`[my link](http://example.com)`);
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile headers', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
# header1
## header2
### header3
#### header4
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should handle same name nested headers', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
# header1
## header2
# header3
## header2
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile ordered list', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
1. foo
2. bar
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile list', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
- foo
- bar
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile tables', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| col 3 is      | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile codespans', () => {
  const compile = marksy({createElement});
  const compiled = compile('install with `$ npm install`')
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile image', () => {
  const compile = marksy({createElement});
  const compiled = compile('![test](http://some.com/image.png)')
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile html', () => {
  const compile = marksy({createElement});
  const compiled = compile(`<div>hello</div>`)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile html as components using babel-core', () => {
  const compile = marksyComponents({createElement, transform: babelCoreTransform});
  const compiled = compile(`<div>hello</div>`)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile html as components using babel-standalone', () => {
  const compile = marksyComponents({createElement, transform: babelStandaloneTransform});
  const compiled = compile(`<div>hello</div>`)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile components using babel-core', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelStandaloneTransform,
    components: {
      Test() {
        return <div>mip</div>
      }
    }
  });
  const compiled = compile(`<Test />`)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile components using babel-standalone', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelCoreTransform,
    components: {
      Test() {
        return <div>mip</div>
      }
    }
  });
  const compiled = compile(`<Test />`)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile components using marksy language using babel-core', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelCoreTransform,
    components: {
      Test() {
        return <div>mip</div>
      }
    }
  });
  const compiled = compile('```marksy\n<Test />\n```')
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile components using marksy language using babel-standalone', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelStandaloneTransform,
    components: {
      Test() {
        return <div>mip</div>
      }
    }
  });
  const compiled = compile('```marksy\n<Test />\n```')
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to compile nested lists', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
- Colors
    - Red
    - Blue
- Shape
  - Triangle
  - Rectangle
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to combine in compilation', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
# hey

- foo

-bar
  `)
  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should produce TOC', () => {
  const compile = marksy({createElement});
  const compiled = compile(`
# foo

## bar

### baz
  `)

  expect(JSON.stringify(compiled.toc, null, 2)).toMatchSnapshot();
});

it('should produce custom tags', () => {
  const compile = marksy({
    createElement,
    elements: {
      h1 (props) {
        return <div>{props.children}</div>
      }
    }
  });
  const compiled = compile(`
# foo
  `)

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should work with Preact', () => {
  const compile = marksy({
    createElement: Preact.h,
    elements: {
      h1 (props) {
        return Preact.h('div', null, props.children)
      }
    }
  });
  const compiled = compile(`
# foo
  `)

  expect(preactRenderToString(Preact.h('div', null, compiled.tree))).toMatchSnapshot();
});

it('should work with Inferno', () => {
  const compile = marksy({
    createElement: infernoCreateElement,
    elements: {
      h1 (props) {
        return infernoCreateElement('div', null, props.children)
      }
    }
  });
  const compiled = compile(`
# foo
  `)

  expect(infernoRenderToString(infernoCreateElement('div', null, compiled.tree))).toMatchSnapshot();
});

it('should allow injecting context to elements', () => {
  const compile = marksy({
    createElement,
    elements: {
      h1 (props) {
        return <div>{props.context.foo}</div>
      }
    }
  });
  const compiled = compile(`
# foo
  `, {}, {
    foo: 'bar'
  })

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should allow injecting context to components using babel-core', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelCoreTransform,
    components: {
      Comp (props) {
        return <div>{props.context.foo}</div>
      }
    }
  });
  const compiled = compile(`
<Comp/>
  `, {}, {
    foo: 'bar'
  })

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should allow injecting context to components using babel-standalone', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelStandaloneTransform,
    components: {
      Comp (props) {
        return <div>{props.context.foo}</div>
      }
    }
  });
  const compiled = compile(`
<Comp/>
  `, {}, {
    foo: 'bar'
  })

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to inline components using babel-core', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelCoreTransform,
    components: {
      Comp (props) {
        return <div>Wuuut</div>
      }
    }
  });
  const compiled = compile(`<p>Hello there <Comp/></p>`)

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should be able to inline components using babel-standalone', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelStandaloneTransform,
    components: {
      Comp (props) {
        return <div>Wuuut</div>
      }
    }
  });
  const compiled = compile(`<p>Hello there <Comp/></p>`)

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should allow overriding code element with components version using babel-core', () => {
  const compile = marksyComponents({
    createElement,
    babelCoreTransform,
    elements: {
      code() {
        return <div>code</div>
      }
    }
  });
  const compiled = compile('Hello `code`');

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should allow overriding code element with components version using babel-standalone', () => {
  const compile = marksyComponents({
    createElement,
    transform: babelStandaloneTransform,
    elements: {
      code() {
        return <div>code</div>
      }
    }
  });
  const compiled = compile('Hello `code`');

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should highlight code with highlight.js', () => {
  const compile = marksy({
    createElement,
    highlight(language, code) {
      return hljs.highlight(language, code).value
    }
  });
  const compiled = compile('```js\nconst foo = "bar"\n```');

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should highlight code with Prism.js', () => {
  const compile = marksy({
    createElement,
    highlight(language, code) {
      return Prism.highlight(code, Prism.languages[language])
    }
  });
  const compiled = compile('```js\nconst foo = "bar"\n```');

  const tree = renderer.create(
    <TestComponent>{compiled.tree}</TestComponent>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
