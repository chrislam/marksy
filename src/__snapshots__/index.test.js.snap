// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow injecting context to components using babel-core 1`] = `
<div>
  <div>
    bar
  </div>
</div>
`;

exports[`should allow injecting context to components using babel-standalone 1`] = `
<div>
  <div>
    bar
  </div>
</div>
`;

exports[`should allow injecting context to elements 1`] = `
<div>
  <div>
    bar
  </div>
</div>
`;

exports[`should allow overriding code element with components version using babel-core 1`] = `
<div>
  <p>
    Hello 
    <div>
      code
    </div>
  </p>
</div>
`;

exports[`should allow overriding code element with components version using babel-standalone 1`] = `
<div>
  <p>
    Hello 
    <div>
      code
    </div>
  </p>
</div>
`;

exports[`should be able to combine in compilation 1`] = `
<div>
  <h1
    id="hey"
  >
    hey
  </h1>
  <ul>
    <li>
      foo
    </li>
  </ul>
  <p>
    -bar
  </p>
</div>
`;

exports[`should be able to compile codespans 1`] = `
<div>
  <p>
    install with 
    <code>
      $ npm install
    </code>
  </p>
</div>
`;

exports[`should be able to compile components using babel-core 1`] = `
<div>
  <div>
    mip
  </div>
</div>
`;

exports[`should be able to compile components using babel-standalone 1`] = `
<div>
  <div>
    mip
  </div>
</div>
`;

exports[`should be able to compile components using marksy language using babel-core 1`] = `
<div>
  <div>
    mip
  </div>
</div>
`;

exports[`should be able to compile components using marksy language using babel-standalone 1`] = `
<div>
  <div>
    mip
  </div>
</div>
`;

exports[`should be able to compile headers 1`] = `
<div>
  <h1
    id="header1"
  >
    header1
  </h1>
  <h2
    id="header1-header2"
  >
    header2
  </h2>
  <h3
    id="header1-header2-header3"
  >
    header3
  </h3>
  <h4
    id="header1-header2-header3-header4"
  >
    header4
  </h4>
</div>
`;

exports[`should be able to compile html 1`] = `
<div>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>hello</div>",
      }
    }
  />
</div>
`;

exports[`should be able to compile html as components using babel-core 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`should be able to compile html as components using babel-standalone 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`should be able to compile image 1`] = `
<div>
  <p>
    <img
      alt="test"
      src="http://some.com/image.png"
    />
  </p>
</div>
`;

exports[`should be able to compile italic text 1`] = `
<div>
  <p>
    hello 
    <em>
      there
    </em>
  </p>
</div>
`;

exports[`should be able to compile links 1`] = `
<div>
  <p>
    <a
      href="http://example.com"
      title={null}
    >
      my link
    </a>
  </p>
</div>
`;

exports[`should be able to compile list 1`] = `
<div>
  <ul>
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ul>
</div>
`;

exports[`should be able to compile nested lists 1`] = `
<div>
  <ul>
    <li>
      Colors
      <ul>
        <li>
          Red
        </li>
        <li>
          Blue
        </li>
      </ul>
    </li>
    <li>
      Shape
      <ul>
        <li>
          Triangle
        </li>
        <li>
          Rectangle
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`should be able to compile ordered list 1`] = `
<div>
  <ol>
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ol>
</div>
`;

exports[`should be able to compile strong text 1`] = `
<div>
  <p>
    hello 
    <strong>
      there
    </strong>
  </p>
</div>
`;

exports[`should be able to compile strong text 2`] = `
<div>
  <p>
    hello 
    <strong>
      there
    </strong>
  </p>
</div>
`;

exports[`should be able to compile tables 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          className={undefined}
        >
          Tables
        </th>
        <th
          className="text-center"
        >
          Are
        </th>
        <th
          className="text-right"
        >
          Cool
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className={undefined}
        >
          col 3 is
        </td>
        <td
          className="text-center"
        >
          right-aligned
        </td>
        <td
          className="text-right"
        >
          $1600
        </td>
      </tr>
      <tr>
        <td
          className={undefined}
        >
          col 2 is
        </td>
        <td
          className="text-center"
        >
          centered
        </td>
        <td
          className="text-right"
        >
          $12
        </td>
      </tr>
      <tr>
        <td
          className={undefined}
        >
          zebra stripes
        </td>
        <td
          className="text-center"
        >
          are neat
        </td>
        <td
          className="text-right"
        >
          $1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should be able to compile text 1`] = `
<div>
  <p>
    hello
  </p>
</div>
`;

exports[`should be able to inline components using babel-core 1`] = `
<div>
  <p>
    Hello there 
    <div>
      Wuuut
    </div>
  </p>
</div>
`;

exports[`should be able to inline components using babel-standalone 1`] = `
<div>
  <p>
    Hello there 
    <div>
      Wuuut
    </div>
  </p>
</div>
`;

exports[`should handle same name nested headers 1`] = `
<div>
  <h1
    id="header1"
  >
    header1
  </h1>
  <h2
    id="header1-header2"
  >
    header2
  </h2>
  <h1
    id="header3"
  >
    header3
  </h1>
  <h2
    id="header3-header2"
  >
    header2
  </h2>
</div>
`;

exports[`should highlight code with Prism.js 1`] = `
<div>
  <pre>
    <code
      className="language-js"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span class=\\"token keyword\\">const</span> foo <span class=\\"token operator\\">=</span> <span class=\\"token string\\">\\"bar\\"</span>",
        }
      }
    />
  </pre>
</div>
`;

exports[`should highlight code with highlight.js 1`] = `
<div>
  <pre>
    <code
      className="language-js"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span class=\\"hljs-keyword\\">const</span> foo = <span class=\\"hljs-string\\">\\"bar\\"</span>",
        }
      }
    />
  </pre>
</div>
`;

exports[`should produce TOC 1`] = `
"[
  {
    \\"id\\": \\"foo\\",
    \\"title\\": \\"foo\\",
    \\"level\\": 1,
    \\"children\\": [
      {
        \\"id\\": \\"foo-bar\\",
        \\"title\\": \\"bar\\",
        \\"level\\": 2,
        \\"children\\": [
          {
            \\"id\\": \\"foo-bar-baz\\",
            \\"title\\": \\"baz\\",
            \\"level\\": 3,
            \\"children\\": []
          }
        ]
      }
    ]
  }
]"
`;

exports[`should produce custom tags 1`] = `
<div>
  <div>
    foo
  </div>
</div>
`;

exports[`should work with Inferno 1`] = `"<div><div>foo</div></div>"`;

exports[`should work with Preact 1`] = `"<div><div>foo</div></div>"`;
